django/__init__.py

```diff
@@ -10,15 +13,36 @@ def setup(set_prefix=True):
     Configure the settings (this happens as a side effect of accessing the
     first setting), configure logging and populate the app registry.
     Set the thread-local urlresolvers script prefix if `set_prefix` is True.
+
+    Raise a ``RuntimeWarning`` when apps interact with the database before
+    the app registry has been fully populated.
     """
     from django.apps import apps
     from django.conf import settings
+    from django.db import connections
     from django.urls import set_script_prefix
     from django.utils.log import configure_logging
 
-    configure_logging(settings.LOGGING_CONFIG, settings.LOGGING)
-    if set_prefix:
-        set_script_prefix(
-            "/" if settings.FORCE_SCRIPT_NAME is None else settings.FORCE_SCRIPT_NAME
-        )
-    apps.populate(settings.INSTALLED_APPS)
+    with ExitStack() as stack:
+        for connection in connections.all():
```

Comment:
This will open all connections, which is quite unexpected (see 4f92cf87b013801810226928ddd20097f6e4fccf) and forces changes to `django/contrib/gis/db/backends/postgis/operations.py`. Can we use only initialized connections?
```suggestion
        for connection in connections.all(initialized_only=True):
```

Reply:
But... then if my `AppConfig` is the first thing that opens a connection, we won't see the warning. So this basically becomes a no-op then? Unless something opened a connection before calling `django.setup()`?

Reply:
Yes. Unfortunately, as far as I'm aware, the current solution is not acceptable as opening all connections during apps initialization can have serious side-effects.

Reply:
I _could_ monkey-patch `django.db.utils.ConnectionHandler.create_connection()` during `django.setup()` though... 🤔

Reply:
I've added a test that checks that `django.setup()` doesn't explicitly initialize database connections on its own. And I've also added tests that check whether the warnings are being raised both when the connections are already initialized and still uninitialized.

I'm checking whether the change to `django/contrib/gis/db/backends/postgis/operations.py` is still necessary...

Reply:
> I'm checking whether the change to `django/contrib/gis/db/backends/postgis/operations.py` is still necessary...

Looks like it isn't. Reverted those changes accordingly 😄

Reply:
You may be able to use the `connection_created` signal to attach the instrumentation: https://adamj.eu/tech/2020/07/23/how-to-make-always-installed-django-database-instrumentation/

Otherwise, I think modifying `CursorWrapper` somewhat could be cleaner than trying to fiddle with database instrumentation.

Reply:
After further thought, I am fairly convinced that a flag in `CursorWrapper` is probably the best way. The cost of checking it per query will be quite small and there will be fewer moving pieces.

Reply:
@adamchainz How about checking `apps.ready` directly? Then we don't need an (undocumented) flag in `CursorWrapper` that basically means the same thing. Plus, I think it communicates more cleanly what we are trying to prevent.

Reply:
Yes, hadn't thought of that, good to avoid duplication.